### ISSUE TITLE
Voice over focus is not as in expected order in FlatList with two columns

### ISSUE THREAD

**Issue by pnikhilp** [facebook/react-native#33714]
### Description

I have a Flatlist with number of columns as 2. When voice over is turned on in iOS device, the first element in the first item of the FlatList is focused and then the first element of the second item in the FlatList is focused, that is in the same row.

Expected- I want to add the focus and accessibility label for all the elements[Text, Button] inside each FlatList item and the voice over focus order should be in a way that it should completely focus/read each element in the first item of the FlatList before the elements in the second item of the FlatList.

This is happening only in iOS that the focus/ reading order is from left to right when voice over is ON. 
Please let me know is there a way to handle the above scenario?
Thank you.

### Version

0.66.3

### Output of `npx react-native info`

System:
    OS: macOS 12.3
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 36.32 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 15.14.0 - ~/Desktop/react-native/node_modules/.bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.8 - ~/.nvm/versions/node/v14.15.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.11.2 - /usr/local/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 21.4, iOS 15.4, macOS 12.3, tvOS 15.4, watchOS 8.5
    Android SDK:
      API Levels: 23, 24, 26, 27, 28, 29
      Build Tools: 26.0.0, 28.0.3, 29.0.2, 29.0.3, 30.0.2
      System Images: android-26 | Intel x86 Atom_64, android-26 | Google APIs Intel x86 Atom, android-27 | Intel x86 Atom_64, android-27 | Google APIs Intel x86 Atom, android-28 | Intel x86 Atom_64, android-29 | Intel x86 Atom_64, android-29 | Google APIs Intel x86 Atom
      Android NDK: Not Found
  IDEs:
    Android Studio: 2021.1 AI-211.7628.21.2111.8309675
    Xcode: 13.3/13E113 - /usr/bin/xcodebuild
  Languages:
    Java: 1.8.0_252 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 17.0.2 => 17.0.2 
    react-native: 0.66.3 => 0.66.3 
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

### Steps to reproduce

Enable voice over and swipe left to right over the elements in a React Native FlatList which is having 2 columns.

**Comment by iscontre** [comment]
Related / dupe of #28299 ?

**Comment by iscontre** [comment]
FWIW, I found the following entry in the react-native-community forum: [Accessibility focus order in React Native](https://github.com/react-native-community/discussions-and-proposals/discussions/389), which has a workaround, but I believe it would be nice to have this as part of the react-native core.

**Comment by blavalla** [comment]
So this is related directly to the way that iOS and Android handle accessibility focus order in native apps, and how it differs from the web.

On the web, focus order is determined by the order of the elements in the DOM. In your case of the two flat lists, as long as those lists were siblings of one another in the DOM, a screen reader would navigate fully through the first list before traversing the second.

But on native mobile, focus order is *not* based on the order of the views in the view hierarchy (at least not *only* based on this), but on their visual position on the screen.  

It's hard to say for sure exactly what Apple's VoiceOver logic is here, since it's closed source, but it seems to take visual position into account much more strongly than Google's Talkback does.

Since Talkback is open source, we can actually look at its exact logic, which is the following:

1. Traverse the view hierarchy starting at the root
2. Move to the element with the lowest Y position (closest to the top of the screen), If multiple elements have identical Y positions look at their X position and pick the one which is lowest (closes to the left).
3. Check if this element is focusable, and if so, add it to a running list of focusable elements.
4. Traverse the branch of the tree from this element down to its children, and repeat steps 2 and 3.
5. Once the bottom of the tree has been reached, go back up to the element with the next lowest Y position, and repeat steps 2 through 4.
6. Once the entire tree has been traversed, you now have a list of all focusable elements, in order.

So you can see that it takes the view hierarchy order into account only *after* it looks at an elements visual position. This means that in your case of two sibling flat lists, in Talkback it would find the top-most-left-most flat list first, and start traversing its descendants before going back up to the second flat list, giving you the desired order.

Since iOS doesn't support the concept of nested accessible elements however, it's system is likely much simpler. Voiceover already knows every element that it needs to focus on, as a user has to explicitly mark each one as an accessible element. Because of this, it can simply look through that list and look at their visual positions alone, rather than having to traverse the entire tree checking if an element should be focusable. This is likely much faster, but does lead to problems like you're describing.

Unfortunately, there's nothing we can do to change VoiceOvers logic here. What we *can* do however is add support for a system to allow users to define manual focus order, as both iOS and Android have ways to handle this. This is no small task however, and would require some pretty core changes to the way that views are created and managed in React Native.

There's an open issue for that feature here - https://github.com/facebook/react-native/issues/30888

Since this is not something that can be fixed directly, and the only way to resolve it is a new API with an issue open for it already, I'm going to close this out.
